### PR TITLE
ARROW-10990: [Rust] Refactor simd comparison kernels to avoid out of bounds reads

### PR DESCRIPTION
**Pull request by jhorstmann**
- [x] Adjust tests so the input data is bigger than one vector lane
- [x] Remove `value_slice` function when ARROW-10989 gets merged

### CODE REVIEW

**Comment by alamb** [comment]
@jhorstmann  -- shall we convert this PR into "ready for review"? It seems important to me but I figured I would wait until you were done with your planned changes before looking into it more carefully.

**Comment by jhorstmann** [comment]
@alamb I finished rewriting the tests and it's ready for review now

File: rust/arrow/src/array/array_primitive.rs [hunk 1]
```diff
@@ -67,19 +67,6 @@ impl<T: ArrowPrimitiveType> PrimitiveArray<T> {
         self.data.is_empty()
     }
 
-    /// Returns a slice for the given offset and length
```
Comment (alamb) [inline rust/arrow/src/array/array_primitive.rs:70]: Since this is a breaking change in the API, it might be worth doing as part of a separate PR. What do you think @jorgecarleitao and @nevi-me ?

File: rust/arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -433,47 +461,74 @@ where
...
+        &remainder_bitmask.to_le_bytes()[0..bit_util::ceil(left_remainder.len(), 8)];
+    result_remainder.copy_from_slice(remainder_mask_as_bytes);
 
     let data = ArrayData::new(
         DataType::Boolean,
-        left.len(),
+        len,
         None,
         null_bit_buffer,
         0,
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/comparison.rs:534]: I see you didn't change this but I don't understand `null_count` here being zero -- it probably can be non-zero as we are just cloning the null buffer from left. In any event we don't need to handle that in this PR

**Review by alamb — APPROVED** [review]
Thank you @jhorstmann! 

I went through this PR carefully and I understand the logic and it all makes sense to me. I think it is worth considering removing `value_slice` as a separate PR but otherwise I think this is ready to go.

cc @nevi-me

File: rust/arrow/src/compute/kernels/comparison.rs (hunk 2 above)
Comment (jhorstmann) [reply rust/arrow/src/compute/kernels/comparison.rs:534]: Good point, no need to recalculate the null count when comparing against a scalar.

File: rust/arrow/src/array/array_primitive.rs (hunk 1 above)
Comment (jhorstmann) [reply rust/arrow/src/array/array_primitive.rs:70]: My rationale was that this would have already been removed as part of ARROW-10989, if not for this one usage in the comparison kernels. The straight-forward replacement would be `values()[offset..offset+len]`.

**Comment by alamb** [comment]
The full set of Rust CI tests did not run on this PR :(

Can you please rebase this PR against [apache/master](https://github.com/apache/arrow) to pick up the changes in https://github.com/apache/arrow/pull/9056 so that they do? 

I apologize for the inconvenience.

File: rust/arrow/src/array/array_primitive.rs (hunk 1 above)
Comment (jorgecarleitao) [reply rust/arrow/src/array/array_primitive.rs:70]: I missed this one, sorry.

I am fine removing it altogether. The change of signature from `safe` to `unsafe` in a previous commit already made this backward incompatible, anyways. As a user, I will prefer to migrate it to `values()[offset..offset+len]` instead of introducing an `unsafe` in my code base.

File: rust/arrow/src/compute/kernels/comparison.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/comparison.rs:537]: This API was deprecated altogether to remove the risk of a wrong count, which I think addresses this altogether :)

**Review by jorgecarleitao — APPROVED** [review]
LGTM. a couple of less `unsafe` 🎉  :)

**Comment by jhorstmann** [comment]
Rebased. The clippy warnings are all in the datafusion logical plan and don't seem related to this PR.

File: rust/arrow/src/array/array_primitive.rs (hunk 1 above)
Comment (jhorstmann) [reply rust/arrow/src/array/array_primitive.rs]: The last merge reintroduced the method, I created a followup ticket to remove it and also from `BooleanArray` so we have the api change documented in a ticket.

Comment (alamb) [reply rust/arrow/src/array/array_primitive.rs]: I see it https://issues.apache.org/jira/browse/ARROW-11099 👍  thanks @jhorstmann
